controllers/usernamespace/controller_test.go

```diff
@@ -0,0 +1,361 @@
+	// the reconciliation needs to run twice for it to be trully finished - we're setting up finalizers etc...
+	if _, err := r.Reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: cheName, Namespace: cheNamespaceName}}); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := r.Reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: cheName, Namespace: cheNamespaceName}}); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func setup(infraType infrastructure.Type, objs ...runtime.Object) (*runtime.Scheme, client.Client, *CheUserNamespaceReconciler) {
+	infrastructure.InitializeForTesting(infraType)
+	devworkspace.CleanCheClusterInstancesForTest()
+	util.IsOpenShift = infraType == infrastructure.OpenShiftv4
+	util.IsOpenShift4 = infraType == infrastructure.OpenShiftv4
+
+	scheme := createTestScheme()
+
+	cl := fake.NewFakeClientWithScheme(scheme, objs...)
+
+	r := &CheUserNamespaceReconciler{
+		client: cl,
+		scheme: scheme,
+		namespaceCache: namespaceCache{
+			client:          cl,
+			knownNamespaces: map[string]namespaceInfo{},
+			lock:            sync.Mutex{},
+		},
+	}
+
+	return scheme, cl, r
+}
+
+func TestSkipsUnlabeledNamespaces(t *testing.T) {
+	test := func(t *testing.T, infraType infrastructure.Type, namespace metav1.Object) {
+		ctx := context.TODO()
+		scheme, cl, r := setup(infraType, namespace.(runtime.Object))
+		setupCheCluster(t, ctx, cl, scheme, "che", "che")
+
+		if _, err := r.Reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: namespace.GetName()}}); err != nil {
+			t.Fatal(err)
+		}
+
+		// no new secret or configmap should be created in the namespace
+		ss := &corev1.SecretList{}
+		if err := cl.List(ctx, ss, client.InNamespace(namespace.GetName())); err != nil {
+			t.Fatal(err)
+		}
+		if len(ss.Items) > 0 {
+			t.Errorf("No secrets expected in the tested namespace but found %d", len(ss.Items))
+		}
+
+		cs := &corev1.ConfigMapList{}
+		if err := cl.List(ctx, cs, client.InNamespace(namespace.GetName())); err != nil {
+			t.Fatal(err)
+		}
+		if len(cs.Items) > 0 {
+			t.Errorf("No configmaps expected in the tested namespace but found %d", len(cs.Items))
+		}
+	}
+
+	t.Run("k8s", func(t *testing.T) {
+		test(t, infrastructure.Kubernetes, &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "ns",
+			},
+		})
+	})
+
+	t.Run("openshift", func(t *testing.T) {
+		test(t, infrastructure.OpenShiftv4, &projectv1.Project{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "prj",
+			},
+		})
+	})
+}
+
+func TestRequiresLabelsToMatchOneOfMultipleCheCluster(t *testing.T) {
+	test := func(t *testing.T, infraType infrastructure.Type, namespace metav1.Object) {
+		ctx := context.TODO()
+		scheme, cl, r := setup(infraType, namespace.(runtime.Object))
+		setupCheCluster(t, ctx, cl, scheme, "che1", "che")
+		setupCheCluster(t, ctx, cl, scheme, "che2", "che")
+
+		res, err := r.Reconcile(reconcile.Request{NamespacedName: types.NamespacedName{Name: namespace.GetName()}})
+		if err != nil {
```

Comment:
these three lines can be replaced with
```suggestion
assert.NilError(t, err)
```